fix(report): coverage of local xsl:variable to use Saxon data alone

galtm opened the pull request:
This pull request starts with the files that @birdya22 attached (thanks!) to issue #2124, to do the following:

- Make code coverage for `xsl:variable` rely only on Saxon trace data for that element, instead of supplementing that data with logic about elements near `xsl:variable`
- Update end-to-end coverage test files that use `xsl:variable`
- <s>Introduce a new end-to-end coverage test file that specifically focuses on instances where `xsl:variable` is marked as "missed" due solely to Saxon optimizations (eba29644a0699d6f67c7c8683379943a76da6f39)</s>

<s>
On that foundation, I added the following:

- Test files updates to reflect behavior of Saxon-HE 12.7, which is the Saxon version this repo uses for running end-to-end coverage tests, and reflect the changes in `external_xsl-package-01.xspec` in #2120 (5b8c9d0ce48ab8fda80114f2ddfe7eecd2607ccf)
- Variant of the optimization-focused test (last item in 1st list above), to make Saxon use `-opt:0`, as discussed in https://github.com/xspec/xspec/issues/2124#issuecomment-2979145091 (a8d5c6978807f1083fb067999a9372f80dd69a68)

Fixes #2124 _except_ the following doc update is still to do:

- [ ] Explain the new logic in the [wiki](https://github.com/xspec/xspec/wiki/Code-Coverage). Suggest that users specify an -opt:... flag as a Saxon custom option, if they want different coverage reporting characteristics.
</s>

galtm committed:
Test updates for xsl:variable using trace data

Files are from birdya22, in ZIP-file linked from https://github.com/xspec/xspec/issues/2124#issuecomment-2987314977

Comments describe behavior with Saxon-HE 12.7 with default optimization

Co-Authored-By: birdya22 <168069809+birdya22@users.noreply.github.com>

galtm committed:
Test focusing on xsl:variable optimization

Files are from birdya22, in ZIP-file linked from https://github.com/xspec/xspec/issues/2124#issuecomment-2983385473

Comments describe behavior with Saxon-HE 12.7 with default optimization

Co-Authored-By: birdya22 <168069809+birdya22@users.noreply.github.com>

galtm committed:
Coverage for xsl:variable to use trace data

Changes are as described by birdya22, in the "Changes Required" section of https://github.com/xspec/xspec/issues/2124#issue-3090450150

Co-Authored-By: birdya22 <168069809+birdya22@users.noreply.github.com>

galtm committed:
Generate expected end-to-end results

galtm committed:
Changes for Saxon-HE 12.7 and PR 2120

galtm committed:
Separate XSpec file with -opt:0

galtm committed:
Regenerate expected end-to-end results

galtm commented on test/end-to-end/cases-coverage/xsl-variable-02-no-opt.xspec:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<?xspec-test enable-coverage?>
+<?xspec-test saxon-custom-options=-opt:0?>
```
With Saxon-HE 12.7, I got the same trace with `-opt:0` or `-opt:-mv`. I don't see a lot of benefit of adding another variant of `xsl-variable-02.xspec` besides this one.

galtm replied:
Update (2026): Removing this -no-opt.xspec file from this PR, so marking this comment as Resolved.

galtm commented on test/end-to-end/cases-coverage/xsl-variable-02.xsl:
```diff
@@ -0,0 +1,67 @@
+      <node type="variable - constant">
+        <xsl:value-of select="$constant02" separator="" />
+      </node>
+      <!-- Local variable inlined with select attribute -->
```
Saxon 12.4 inlined the variables in this section of the file, but Saxon 12.7 didn't. @birdya22 , do you have ideas about modifying this section, or updating the comments on lines 52 and 57?

birdya22 replied:
I will include this at part of my Saxon 12.7 changes.

galtm replied:
I hadn't realized you were planning to deliver more changes. Sounds good. Thanks!

galtm replied:
Marking this comment as Resolved due to updates in #2124 for Saxon 12.9.

galtm commented on test/end-to-end/cases-coverage/xsl-variable-02.xsl:
```diff
@@ -0,0 +1,67 @@
+        </xsl:variable>
+        <xsl:sequence select="$trivial02" />
+      </node>
+      <!-- Local variable constant with select attribute -->
```
The Saxon 12.7 output from `-explain` didn't say it optimized a constant. Instead, it said `OPT : Eliminated unused variable constant01` and `OPT : Eliminated unused variable constant02`. The variables aren't unused, so maybe Saxon inlined them and _then_ found that the original declarations were unused. Should I push a new commit that changes the "optim constant" instances in this file to reflect Saxon 12.7 behavior?

birdya22 replied:
I was intending to include this at part of my Saxon 12.7 changes.

galtm replied:
Update (2026): This PR no longer pertains to optimizations, so marking this comment as Resolved.

galtm commented on test/end-to-end/cases-coverage/expected/stylesheet/xsl-variable-02-no-opt-coverage.html:
```diff
@@ -0,0 +1,122 @@
+44: <span class="comment">           Needed as attribute to make it constant --&gt;</span>
+45: <span class="whitespace">      </span><span class="hit">&lt;xsl:variable name="constant02" as="xs:string+"&gt;</span><span class="whitespace">                         </span><span class="comment">&lt;!-- Expected miss (optim constant) --&gt;</span>
+46: <span class="whitespace">        </span><span class="hit">&lt;xsl:text&gt;</span><span class="hit">2</span><span class="hit">&lt;/xsl:text&gt;</span><span class="whitespace">                                                 </span><span class="comment">&lt;!-- Expected miss (optim constant) --&gt;</span>
+47: <span class="whitespace">        </span><span class="missed">&lt;xsl:text&gt;</span><span class="missed">00</span><span class="missed">&lt;/xsl:text&gt;</span><span class="whitespace">                                                </span><span class="comment">&lt;!-- Expected miss (optim constant) --&gt;</span>
```
I was surprised that with optimizations off, line 47 was missed. Line 46 was traced with `class="net.sf.saxon.expr.Literal"`, so why not line 47, too?

galtm replied:
Update (2026): Removing this -no-opt.xspec file from this PR, so marking this comment as Resolved.

galtm commented on test/end-to-end/cases-coverage/xsl-variable-02-no-opt.xspec:
```diff
@@ -0,0 +1,9 @@
+<x:description xmlns:x="http://www.jenitennison.com/xslt/xspec"
+               stylesheet="xsl-variable-02.xsl">
+   <!-- When executing this test with the Saxon -opt:0 setting, the
+      "Expected miss" lines in xsl-variable-02.xsl should be hit. -->
```
This comment is not quite clear, given that line 47 doesn't get traced or marked "hit", contrary to my expectation.

birdya22 replied:
I need to take a bit more time to look at this.

birdya22 replied:
In 12.7 the Saxon trace shows that it is doing a `fn:string-join()` internally, so I expect it is basically treating the 2 literals as a single entity and only traces the first one. The `explain` output shows something similar.
I would need to have a play to see what could replace the second element, so that it is traced, while keeping the overall value as a `constant` (that will be some way down my priority list at the moment).
In 12.4 neither lines 46 or 47 are hit (not sure why this is).

galtm replied:
Update (2026): This PR no longer contains this file, so marking this comment as Resolved.

birdya22 reviewed (commented):
It's taken me a very long time to review this, sorry.
Some of the changes are no longer needed with Saxon 12.9 and, yes, I probably got a bit carried away with optimization.
I would like to go back to the original issue and update it for Saxon 12.9, without the optimization changes (which I will think about separately).

galtm committed:
Merge branch 'main' into coverage-xsl-variable

galtm committed:
Files from birdya22 in ZIP-file

Attached June 13, 2026 to
https://github.com/xspec/xspec/issues/2124#issuecomment-4697810736

Co-Authored-By: birdya22 <168069809+birdya22@users.noreply.github.com>

galtm committed:
Editorial tweak in Markdown document

galtm committed:
Mark xsl:import, xsl:mode as expected ignored

galtm committed:
Remove no-optimization test case
